ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,467 @@
+		return err
+	}
+	return pm.setUnattendedModeAsConfigured()
+}
+
+func (pm *ProfileManager) writePrefsToStore(key ipn.StateKey, prefs ipn.PrefsView) error {
+	if key == "" {
+		return nil
+	}
+	if err := pm.store.WriteState(key, prefs.ToBytes()); err != nil {
+		pm.logf("WriteState(%q): %v", key, err)
+		return err
+	}
+	return nil
+}
+
+// Profiles returns the list of known profiles.
+func (pm *ProfileManager) Profiles() []ipn.LoginProfile {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	var profiles []ipn.LoginProfile
+	for _, p := range pm.knownProfiles {
+		if p.LocalUserID == pm.currentUserID {
+			profiles = append(profiles, *p)
+		}
+	}
+	slices.SortFunc(profiles, func(a, b ipn.LoginProfile) bool {
+		return a.Name < b.Name
+	})
+	return profiles
+}
+
+// SwitchProfile switches to the profile with the given name.
+func (pm *ProfileManager) SwitchProfile(profile ipn.ProfileID) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	kp, ok := pm.knownProfiles[profile]
+	if !ok {
+		return ErrProfileNotFound
+	}
+
+	if pm.currentProfile != nil && kp.Key == pm.currentProfile.Key && pm.prefs.Valid() {
+		return nil
+	}
+	if kp.LocalUserID != pm.currentUserID {
+		return fmt.Errorf("profile %q is not owned by current user", profile)
+	}
+	prefs, err := pm.loadSavedPrefs(kp.Key)
+	if err != nil {
+		return err
+	}
+	pm.prefs = prefs
+	pm.currentProfile = kp
+	pm.isNewProfile = false
+	return pm.setAsUserSelectedProfileLocked()
+}
+
+func (pm *ProfileManager) setAsUserSelectedProfileLocked() error {
+	k := ipn.CurrentProfileKey(pm.currentUserID)
+	if pm.currentProfile == nil {
+		return pm.store.WriteState(k, nil)
+	}
+	return pm.store.WriteState(k, []byte(pm.currentProfile.Key))
+}
+
+func (pm *ProfileManager) loadSavedPrefs(key ipn.StateKey) (ipn.PrefsView, error) {
+	bs, err := pm.store.ReadState(key)
+	if err != nil {
+		if err == ipn.ErrStateNotExist {
+			return emptyPrefs, nil
+		}
+		return ipn.PrefsView{}, err
+	}
+	savedPrefs, err := ipn.PrefsFromBytes(bs)
+	if err != nil {
+		return ipn.PrefsView{}, fmt.Errorf("PrefsFromBytes: %v", err)
+	}
+	pm.logf("using backend prefs for %q: %v", key, savedPrefs.Pretty())
+
+	// Ignore any old stored preferences for https://login.tailscale.com
+	// as the control server that would override the new default of
+	// controlplane.tailscale.com.
+	if savedPrefs.ControlURL != "" &&
+		savedPrefs.ControlURL != ipn.DefaultControlURL &&
+		ipn.IsLoginServerSynonym(savedPrefs.ControlURL) {
+		savedPrefs.ControlURL = ""
+	}
+	return savedPrefs.View(), nil
+}
+
+// CurrentProfile returns the name and ID of the current profile, or "" if the profile
+// is not named.
+func (pm *ProfileManager) CurrentProfile() ipn.LoginProfile {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if pm.currentProfile == nil {
+		return ipn.LoginProfile{}
+	}
+	return *pm.currentProfile
+}
+
+var ErrProfileNotFound = errors.New("profile not found")
+
+// DeleteProfile removes the profile with the given id. It is a no-op if the
```

Comment:
It does return an error though, so a caller may end up stopping instead of continuing if the profile does not exist.

Reply:
Updated docs.